### PR TITLE
Widget manager (1/3) add redux to handle state change in a predictable way

### PR DESCRIPTION
**Pull request by sandinosaso**
**Description**

We want to handle the widgets-manager library state. This will help us have a centralized state that makes our library to be more predictable and to behave consistently.

**Acceptance Criteria**

- [x] State is only modified via reducers
- [x] Loaded apps are persisted in the state
- [x] Mounted / Unmounted apps are persisted in the state
- [x] Current user session is persisted in the state

**Commercial Value**

It will help us to have apps that can not be opened twice if we want to (for example because user closed the app).

**Customer Value**

Customers will not be prompted many times after closing an app. We can remember user state so we can launch the same apps he was viewing in the last session.

**Help screen shoots**

This is how the state shape is (just plain object with 3 main keys). Each reducer is in charge of updating one part of the state

Screen Shot 2019-10-29 at 1 56 59 AM

Each of those apps are just plain objects that represents our apps (same as ever):
Screen Shot 2019-10-29 at 1 56 43 AM

### CODE REVIEW

File: packages/widgets-manager/src/reducers/__tests__/mounted_apps.spec.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import {
+  syncData,
+  mountAppSuccess,
+  unMountAppSuccess,
+} from '../../store/actions';
+
+describe('When it recieves syncData action', () => {
+  it('Should set the received mounted_apps', () => {
+    const mockMountedApps = {
+      1: true,
```
Comment (PGrothaus) [inline packages/widgets-manager/src/reducers/__tests__/mounted_apps.spec.ts]: Do we always identify apps by an integer? Do we have a mapping to the app-name somewhere?

File: packages/widgets-manager/src/reducers/__tests__/user.spec.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    const stateAfterReducer = userReducer(initialState, actionToTest);
+
+    expect(stateAfterReducer).toEqual({
+      uid: 'uid123',
+      isAnonymous: true,
+      metadata: { vip: true },
+    });
+  });
+});
```
Comment (PGrothaus) [inline packages/widgets-manager/src/reducers/__tests__/user.spec.ts]: Do we want to add a test that updates a state? In this test the initial state is empty, no?

File: packages/widgets-manager/src/reducers/user.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * @param state: current application state
+ * @param action: the current action fired in the application
+ * @returns state: the new application state
+ */
+const reducer: any = (
+  prevState: any = initialState,
+  action: Action
+): any => {
+  switch (action.type) {
+    case ActionType.UPDATE_USER_DATA:
```
Comment (alex-sysoev) [inline packages/widgets-manager/src/reducers/user.ts]: When you update user data and save it to storage ? What about using the same library that uses widget but configure it with fire base?

Comment (sandinosaso) [reply packages/widgets-manager/src/reducers/user.ts]: I thought about that and event investigated a little, and found out that to make redux-persist work with firebase there is a need to use react-redux-firebase library for bindings. Using it directly without React is not possible. So because widget-manager is plain vanilla JS I did not want to add react to it just to use redux-persist.

I thought about it and maybe in the future, if this project grows a lot we may want to add a lightweight version of React there are some options and so use it (https://github.com/preactjs/preact)
For now, we just do it without libraries.

Comment (sandinosaso) [reply packages/widgets-manager/src/reducers/user.ts]: Created an issue for that: https://github.com/lets-talk/chat-utils/issues/69
